pkg/workloads/reservation.go

```diff
@@ -147,35 +149,27 @@ func (a *API) create(r *http.Request) (interface{}, mw.Response) {
 		if err != nil {
 			return nil, mw.Error(err, http.StatusInternalServerError)
 		}
+		// FreeToUse allows payment with FreeTFT
 		if node.FreeToUse {
 			freeNodes++
+			paidNodes++
 		} else {
 			paidNodes++
 		}
 	}
 
-	// don't allow mixed nodes
-	if freeNodes > 0 && paidNodes > 0 {
-		return nil, mw.Error(errors.New("reservation can only contain either free nodes or paid nodes, not both"), http.StatusBadRequest)
-	}
-
+	acceptFreeTFT := paidNodes == 0 && freeNodes > 0
 	currencies := []string{}
 	// filter out FreeTFT if its a paid reservation
-	if paidNodes > 0 {
+	if !acceptFreeTFT {
 		for _, c := range reservation.DataReservation.Currencies {
 			if c != freeTFT {
 				currencies = append(currencies, c)
 			}
 		}
-	}
-
-	// filter out anything but FreeTFT for a free reservation
-	if freeNodes > 0 {
-		for _, c := range reservation.DataReservation.Currencies {
-			if c == freeTFT {
-				currencies = append(currencies, c)
-			}
-		}
+	} else {
+		//if it's a free reservation only allow freeTFT
+		currencies = []string{freeTFT}
```

Comment:
This will change the behavior of the reservation. Previously a client needed to explicitly inform the system that he wants to pay in FreeTFT, now we will just set the currency for the reservation to FreeTFT in this case. Is this intended?

Reply:
My intend was to have some kind of negotiation of  the currency to use.
Since the user send a list of usable currency and based on the list of node used we also have a list of usable currency. The explorer should pick the first currency  that is proposed by both party.
Does this change prevent this now ?

Reply:
Well, currency negotiation as you describe already happens. But it happens in the escrow. This code here filters out currencies which are unacceptable for reservations. In other words, a paid resesrvation filtered out `FreeTFT`, which is only usefull for making free reservations, and a free reservation filtered out everything but `FreeTFT`. But in any case, the user needed to explicitly set the currency he used. Should there no longer be a curreny after this step (i.e. he tried to make a paid reservation and only wanted to pay `FreeTFT`, or the other way round), then the currency list is empty when it gets to the escrow, which will fail to find a match between the currencies offered and the currencies accepted (as there is no longer a currency offered), and return an appropriate error. 

Right now, you will always set `FreeTFT` as an acceptable currency by the client, if the reservation is for free nodes. This means that, if I make a "free" reservation, but I set the currency field ot [`TFT`], this code will modify that currency field to [`FreeTFT`] regardless, which is a currency I did not offer. This goes through to the escrow, the escrow will find that indeed all nodes accept `FreeTFT`, and return payment information to the client. While this payment information does specify that it needs `FreeTFT`, this might not be obvious to a client, who might end up paying regular `TFT` since that is what he specified, and be surprised because his reservations did not succeed

In the old code, if a user wanted a free reservation, he needed to also set `FreeTFT` as the currency to use. The other currencies would be filtered out because deploying happened on free nodes. The

Reply:
ok so my logic is wrong and I'm not sure what I need to do to actually implement what is described in https://github.com/threefoldtech/tfexplorer/issues/50.
From what I understand from your comment everything was already in place for it to work they way we want. But why then was it not working like that already now ?

Reply:
Currency conversion yes. The thing is, the old code had some logic which forced a user to deploy either on all free nodes, or no free nodes. So that needs to go. Furthermore, I think this else branch for the currency (which forces FreeTFT) can go as well. Instead we can just count the amount of nodes which are marked as free to use. If all nodes are free to use, we also allow FreeTFT, but if they are not, we filter out FreeTFT (this already happens in the primary if branch). Then its up to the user to decide. If he wants to pay in FreeTFT, he'll have to set that as the only currency, or the first currency to use. If he wants to pay for the reservation, but keep freeTFT as an option in case a farmer only accepts TFTA and another one only accepts TFT, but all node are free to use, that is also an option